tests/infra/ccf.py

```diff
@@ -123,46 +116,49 @@ def start_and_join(self, args):
 
         LOG.info("Starting nodes on {}".format(hosts))
 
+        self.add_members([1, 2, 3])
+        self.add_users([1, 2, 3])
+
+        if args.app_script:
+            infra.proc.ccall("cp", args.app_script, args.build_dir).check_returncode()
+        if args.gov_script:
+            infra.proc.ccall("cp", args.gov_script, args.build_dir).check_returncode()
+        LOG.info("Lua scripts copied")
+
+        primary = None
         for i, node in enumerate(self.nodes):
             dict_args = vars(args)
             forwarded_args = {
                 arg: dict_args[arg] for arg in Network.node_args_to_forward
             }
             try:
+                primary, _ = self.find_primary() if i != 0 else (None, None)
                 node.start(
+                    infra.remote.StartType.start
+                    if i == 0
+                    else infra.remote.StartType.join,
                     lib_name=args.package,
                     node_status=node_status[i],
                     workspace=args.workspace,
                     label=args.label,
-                    other_quote=None,
-                    other_quoted_data=None,
+                    target_rpc_address=f"{primary.host}:{primary.rpc_port}"
+                    if primary
+                    else None,
+                    members_certs="member*_cert.pem" if i == 0 else None,
+                    users_certs="user*_cert.pem" if i == 0 else None,
                     **forwarded_args,
                 )
+                node.network_state = NodeNetworkState.joined
             except Exception:
                 LOG.exception("Failed to start node {}".format(i))
                 raise
-        LOG.info("All remotes started")
-
-        if args.app_script:
-            infra.proc.ccall("cp", args.app_script, args.build_dir).check_returncode()
-        if args.gov_script:
-            infra.proc.ccall("cp", args.gov_script, args.build_dir).check_returncode()
-        LOG.info("Lua scripts copied")
-
-        self.nodes_json()
-        self.add_members([1, 2, 3])
-        self.add_users([1, 2, 3])
-        self.genesis_generator(args)
-
-        primary = self.nodes[0]
-        primary.start_network(self)
-
-        for node in self.nodes[1:]:
-            node.join_network(self)
 
-        primary.network_state = NodeNetworkState.joined
+        self.wait_for_node_commit_sync()
```

Comment:
This is unsafe as not all the nodes could be part of the network (i.e. global commit = 0). Instead, we should check that all nodes have joined the network by calling `getCommit` on all nodes with the user frontend.